R/class-spec.R

```diff
@@ -92,8 +92,45 @@ class_constructor <- function(.x, ...) {
     stop(sprintf("Can't construct %s", class_friendly(.x)), call. = FALSE)
   )
 }
+
+class_construct_expr <- function(.x, ...) {
+  f <- class_constructor(.x)
+  # If the constructor is a closure wrapping a simple expression, try to extract the expression
+  # (mostly for nicer printing and introspection.)
+  fb <- body(f)
+  ff <- formals(f)
+  fe <- environment(f)
+
+  # early return if not safe to unwrap
+  if(length(list(...)))
+    return(as.call(list(f, ...)))
+
+  if(!identical(fe, baseenv()))
+    return(as.call(list(f, ...)))
+
+  if("new_object" %in% all.names(fb))
```

Comment:
It would be nice to see some explicit tests for these cases.

Reply:
`class_construct()` is already quite thoroughly tested in https://github.com/RConsortium/S7/blob/main/tests/testthat/test-class-spec.R (that's the reason we have so many branches here 😃)

Reply:
Ok, great.